File: app/src/main/java/me/tagavari/airmessage/messaging/viewbinder/VBConversation.java
package me.tagavari.airmessage.messaging.viewbinder;

import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.util.Pair;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;

import java.util.List;
import java.util.Optional;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Observable;
import me.tagavari.airmessage.MainApplication;
import me.tagavari.airmessage.R;
import me.tagavari.airmessage.helper.ContactHelper;
import me.tagavari.airmessage.helper.ConversationBuildHelper;
import me.tagavari.airmessage.helper.LanguageHelper;
import me.tagavari.airmessage.helper.ResourceHelper;
import me.tagavari.airmessage.messaging.ConversationInfo;
import me.tagavari.airmessage.messaging.ConversationPreview;
import me.tagavari.airmessage.messaging.MemberInfo;
import me.tagavari.airmessage.util.IndexedItem;

/**
 * Binds conversation data to a view holder
 */
public class VBConversation {
	private static final String TAG = VBConversation.class.getSimpleName();
	
	private static final int maxUsersToDisplay = 4;
	
	/**
	 * Binds a conversation's members to an icon group
	 * @param context The context to use
	 * @param iconGroup A view representing a group of members
	 * @param conversationInfo The conversation to display
	 * @return A completable representing this task's state
	 */
	public static Completable bindUsers(Context context, ViewGroup iconGroup, ConversationInfo conversationInfo) {
		return bindUsers(context, iconGroup, conversationInfo.getMembers());
	}
	
	/**
	 * Binds a list of conversation members to an icon group
	 * @param context The context to use
	 * @param iconGroup A view representing a group of members
	 * @param members A list of members to display
	 * @return A completable representing this task's state
	 */
	public static Completable bindUsers(Context context, ViewGroup iconGroup, List<MemberInfo> members) {
		if(members.isEmpty()) {
			//Hide all views
			for(int i = 0; i < maxUsersToDisplay; i++) {
				View child = iconGroup.getChildAt(i);
				if(child instanceof ViewGroup) child.setVisibility(View.GONE);
			}
			
			return Completable.complete();
		}
		
		//Getting the view data
		int usersToDisplay = Math.min(members.size(), maxUsersToDisplay);
		int viewIndex = usersToDisplay - 1;
		View viewAtIndex = iconGroup.getChildAt(viewIndex);
		
		//Retrieving the target view, inflating if necessary
		ViewGroup iconView;
		if(viewAtIndex instanceof ViewStub) iconView = (ViewGroup) ((ViewStub) viewAtIndex).inflate();
		else {
			iconView = (ViewGroup) viewAtIndex;
			iconView.setVisibility(View.VISIBLE);
		}
		
		//Hiding all other views
		for(int i = 0; i < maxUsersToDisplay; i++) {
			if(i == viewIndex) continue;
			View child = iconGroup.getChildAt(i);
			if(child instanceof ViewGroup) child.setVisibility(View.GONE);
		}
		
		//Getting user data for each member
		return Observable.range(0, usersToDisplay)
				//Map to an object containing each member and their index
				.map((index) -> new IndexedItem<>(index, members.get(index)))
				.doOnNext((data) -> {
					//Getting the child view
					View child = iconView.getChildAt(data.index);
					
					//Getting the views
					ImageView imageDefault = child.findViewById(R.id.profile_default);
					ImageView imageProfile = child.findViewById(R.id.profile_image);
					
					//Setting the default profile tint
					imageDefault.setVisibility(View.VISIBLE);
					imageDefault.setColorFilter(data.item.getColor(), PorterDuff.Mode.MULTIPLY);
					
					//Resetting the contact image
					imageProfile.setImageBitmap(null);
				})
				//Get user info for each member
				.flatMapSingle((data) ->
						MainApplication.getInstance().getUserCacheHelper().getUserInfo(context, data.item.getAddress())
								.map(Optional::of)
								.onErrorReturnItem(Optional.empty())
								.map((optionalUserInfo) -> new IndexedItem<>(data.index, new Pair<>(data.item, optionalUserInfo.orElse(null)))))
				.doOnNext((data) -> {
					//Getting the child view
					View child = iconView.getChildAt(data.index);
					
					//Getting the views
					ImageView imageDefault = child.findViewById(R.id.profile_default);
					ImageView imageProfile = child.findViewById(R.id.profile_image);
					
					//Checking if a user was found
					if(data.item.second != null) {
						//Loading the user's picture
						Glide.with(context)
								.load(ContactHelper.getContactImageURI(data.item.second.getContactID()))
								.listener(new RequestListener<Drawable>() {
									@Override
									public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
										return false;
									}
									
									@Override
									public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
										//Swapping to the profile view
										imageDefault.setVisibility(View.GONE);
										imageProfile.setVisibility(View.VISIBLE);
										
										return false;
									}
								})
								.into(imageProfile);
					}
				})
				//We're done
				.ignoreElements();
	}
	
	/**
	 * Binds a conversation's title to a TextView
	 * @param context The context to use
	 * @param label The TextView to use
	 * @param conversationInfo The conversation to display
	 * @return A completable representing this task's state
	 */
	public static Completable bindTitle(Context context, TextView label, ConversationInfo conversationInfo) {
		//Apply a temporary title right away
		label.setText(ConversationBuildHelper.buildConversationTitleDirect(context, conversationInfo));
		
		//Apply a title with member names asynchronously
		return ConversationBuildHelper.buildConversationTitle(context, conversationInfo).doOnSuccess(label::setText).ignoreElement();
	}
	
	/**
	 * Binds a conversation's unread status
	 * @param context The context to use
	 * @param labelTitle The TextView of the conversation's title
	 * @param labelMessage The TextView of the conversation's preview message
	 * @param labelUnread The TextView of the conversation's unread indicator
	 * @param unreadCount The amount of unread messages; set to 0 for none
	 */
	public static void bindUnreadStatus(Context context, TextView labelTitle, TextView labelMessage, TextView labelUnread, int unreadCount) {
		if(unreadCount > 0) {
			labelTitle.setTypeface(null, Typeface.BOLD);
			labelTitle.setTextColor(context.getResources().getColor(R.color.colorPrimary, null));
			
			labelMessage.setTypeface(null, Typeface.BOLD);
			labelMessage.setTextColor(ResourceHelper.resolveColorAttr(context, android.R.attr.textColorPrimary));
			
			labelUnread.setVisibility(View.VISIBLE);
			labelUnread.setText(LanguageHelper.intToFormattedString(context.getResources(), unreadCount));
		} else {
			labelTitle.setTypeface(null, Typeface.NORMAL);
			labelTitle.setTextColor(ResourceHelper.resolveColorAttr(context, android.R.attr.textColorPrimary));
			
			labelMessage.setTypeface(null, Typeface.NORMAL);
			labelMessage.setTextColor(ResourceHelper.resolveColorAttr(context, android.R.attr.textColorSecondary));
			
			labelUnread.setVisibility(View.GONE);
		}
	}
	
	/**
	 * Binds a conversation's preview section
	 * @param context The context to use
	 * @param labelMessage The TextView of the conversation's preview message
	 * @param labelStatus The TextView of the conversation's status
	 * @param preview The preview item to display
	 */
	public static void bindPreview(Context context, TextView labelMessage, TextView labelStatus, @Nullable ConversationPreview preview) {
		if(preview == null) {
			//Set everything to "unknown"
			labelMessage.setText(R.string.part_unknown);
			labelStatus.setText(R.string.part_unknown);
			labelStatus.setTextColor(ResourceHelper.resolveColorAttr(context, android.R.attr.textColorSecondary));
			
			return;
		}
		
		//Setting the message preview
		labelMessage.setText(preview.buildString(context));
		
		if(preview instanceof ConversationPreview.Message && ((ConversationPreview.Message) preview).isError()) {
			//Setting the status to "not sent"
			labelStatus.setText(R.string.message_senderror);
			labelStatus.setTextColor(context.getResources().getColor(R.color.colorError, null));
		} else {
			//Setting the status to the preview time
			labelStatus.setText(LanguageHelper.getLastUpdateStatusTime(context, preview.getDate()));
			labelStatus.setTextColor(ResourceHelper.resolveColorAttr(context, android.R.attr.textColorSecondary));
		}
		
	}
	
	/**
	 * Binds a conversation's selected indicator
	 * @param mainView The view of the conversation
	 * @param iconGroup A view representing the conversation's group of members
	 * @param selectionIndicator The conversation's selection indicator view
	 * @param isSelected Whether this conversation is selected
	 * @param animate Whether to animate this change
	 */
	public static void bindSelectionIndicator(View mainView, ViewGroup iconGroup, View selectionIndicator, View selectionTint, boolean isSelected, boolean animate) {
		mainView.setSelected(isSelected);
		
		if(animate) {
			if(isSelected) {
				iconGroup.animate().alpha(0).withEndAction(() -> iconGroup.setVisibility(View.INVISIBLE));
				selectionIndicator.animate().alpha(1).withStartAction(() -> selectionIndicator.setVisibility(View.VISIBLE));
				selectionTint.animate().alpha(1).withStartAction(() -> selectionTint.setVisibility(View.VISIBLE));
			} else {
				iconGroup.animate().alpha(1).withStartAction(() -> iconGroup.setVisibility(View.VISIBLE));
				selectionIndicator.animate().alpha(0).withEndAction(() -> selectionIndicator.setVisibility(View.GONE));
				selectionTint.animate().alpha(0).withEndAction(() -> selectionTint.setVisibility(View.GONE));
			}
		} else {
			if(isSelected) {
				iconGroup.setVisibility(View.INVISIBLE);
				iconGroup.setAlpha(0);
				
				selectionIndicator.setVisibility(View.VISIBLE);
				selectionIndicator.setAlpha(1);
				selectionTint.setVisibility(View.VISIBLE);
				selectionTint.setAlpha(1);
			} else {
				iconGroup.setVisibility(View.VISIBLE);
				iconGroup.setAlpha(1);
				
				selectionIndicator.setVisibility(View.GONE);
				selectionIndicator.setAlpha(0);
				selectionTint.setVisibility(View.GONE);
				selectionTint.setAlpha(0);
			}
		}
	}
}

File: app/src/main/java/me/tagavari/airmessage/messaging/viewbinder/VBConversation.kt
package me.tagavari.airmessage.messaging.viewbinder

import android.content.Context
import android.graphics.PorterDuff
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.view.View
import android.view.ViewGroup
import android.view.ViewStub
import android.widget.ImageView
import android.widget.TextView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import me.tagavari.airmessage.MainApplication
import me.tagavari.airmessage.R
import me.tagavari.airmessage.common.data.UserCacheHelper
import me.tagavari.airmessage.common.helper.ConversationBuildHelper.buildConversationTitle
import me.tagavari.airmessage.common.helper.ConversationBuildHelper.buildConversationTitleDirect
import me.tagavari.airmessage.common.helper.LanguageHelper.getLastUpdateStatusTime
import me.tagavari.airmessage.common.helper.LanguageHelper.intToFormattedString
import me.tagavari.airmessage.common.helper.ResourceHelper.resolveColorAttr
import me.tagavari.airmessage.common.messaging.ConversationInfo
import me.tagavari.airmessage.common.messaging.ConversationPreview
import me.tagavari.airmessage.common.messaging.MemberInfo
import me.tagavari.airmessage.common.util.IndexedItem
import java.util.*

/**
 * Binds conversation data to a view holder
 */
object VBConversation {
	private val TAG = VBConversation::class.java.simpleName
	private const val maxUsersToDisplay = 4
	
	/**
	 * Binds a conversation's members to an icon group
	 * @param context The context to use
	 * @param iconGroup A view representing a group of members
	 * @param conversationInfo The conversation to display
	 * @return A completable representing this task's state
	 */
	@JvmStatic
	fun bindUsers(context: Context, iconGroup: ViewGroup, conversationInfo: ConversationInfo): Completable {
		return bindUsers(context, iconGroup, conversationInfo.members)
	}
	
	/**
	 * Binds a list of conversation members to an icon group
	 * @param context The context to use
	 * @param iconGroup A view representing a group of members
	 * @param members A list of members to display
	 * @return A completable representing this task's state
	 */
	@JvmStatic
	fun bindUsers(context: Context, iconGroup: ViewGroup, members: List<MemberInfo>): Completable {
		if(members.isEmpty()) {
			//Hide all views
			for(i in 0 until maxUsersToDisplay) {
				val child = iconGroup.getChildAt(i)
				if(child is ViewGroup) child.visibility = View.GONE
			}
			
			return Completable.complete()
		}
		
		//Getting the view data
		val usersToDisplay = members.size.coerceAtMost(maxUsersToDisplay)
		val viewIndex = usersToDisplay - 1
		val viewAtIndex = iconGroup.getChildAt(viewIndex)
		
		//Retrieving the target view, inflating if necessary
		val iconView: ViewGroup
		if(viewAtIndex is ViewStub) {
			iconView = viewAtIndex.inflate() as ViewGroup
		} else {
			iconView = viewAtIndex as ViewGroup
			iconView.visibility = View.VISIBLE
		}
		
		//Hiding all other views
		for(i in 0 until maxUsersToDisplay) {
			if(i == viewIndex) continue
			
			val child = iconGroup.getChildAt(i)
			if(child is ViewGroup) child.visibility = View.GONE
		}
		
		//Getting user data for each member
		return Observable.range(0, usersToDisplay)
			//Map to an object containing each member and their index
			.map { i -> IndexedItem(i, members[i]) }
			.doOnNext { data: IndexedItem<MemberInfo> ->
				//Getting the child view
				val child = iconView.getChildAt(data.index)
				
				//Getting the views
				val imageDefault = child.findViewById<ImageView>(R.id.profile_default)
				val imageProfile = child.findViewById<ImageView>(R.id.profile_image)
				
				//Setting the default profile tint
				imageDefault.visibility = View.VISIBLE
				imageDefault.setColorFilter(data.item.color, PorterDuff.Mode.MULTIPLY)
				
				//Resetting the contact image
				imageProfile.setImageBitmap(null)
			}
			//Get user info for each member
			.flatMapSingle { data: IndexedItem<MemberInfo> ->
				MainApplication.instance.userCacheHelper.getUserInfo(context, data.item.address)
					.map { Optional.of(it) }
					.onErrorReturnItem(Optional.empty())
					.map { optionalUserInfo -> IndexedItem(data.index, Pair(data.item, optionalUserInfo.orElse(null))) }
			}
			.doOnNext { data: IndexedItem<Pair<MemberInfo, UserCacheHelper.UserInfo?>> ->
				//Getting the child view
				val child = iconView.getChildAt(data.index)
				
				//Getting the views
				val imageDefault = child.findViewById<ImageView>(R.id.profile_default)
				val imageProfile = child.findViewById<ImageView>(R.id.profile_image)
				
				//Checking if a user was found
				if(data.item.second != null) {
					//Loading the user's picture
					Glide.with(context)
						.load(data.item.second!!.thumbnailURI)
						.listener(object : RequestListener<Drawable?> {
							override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<Drawable?>, isFirstResource: Boolean): Boolean {
								return false
							}
							
							override fun onResourceReady(resource: Drawable?, model: Any?, target: Target<Drawable?>, dataSource: DataSource, isFirstResource: Boolean): Boolean {
								//Swapping to the profile view
								imageDefault.visibility = View.GONE
								imageProfile.visibility = View.VISIBLE
								return false
							}
						})
						.into(imageProfile)
				}
			} //We're done
			.ignoreElements()
	}
	
	/**
	 * Binds a conversation's title to a TextView
	 * @param context The context to use
	 * @param label The TextView to use
	 * @param conversationInfo The conversation to display
	 * @return A completable representing this task's state
	 */
	@JvmStatic
	fun bindTitle(context: Context, label: TextView, conversationInfo: ConversationInfo): Completable {
		//Apply a temporary title right away
		label.text = buildConversationTitleDirect(context, conversationInfo)
		
		//Apply a title with member names asynchronously
		return buildConversationTitle(context, conversationInfo)
			.doOnSuccess { label.text = it }
			.ignoreElement()
	}
	
	/**
	 * Binds a conversation's unread status
	 * @param context The context to use
	 * @param labelTitle The TextView of the conversation's title
	 * @param labelMessage The TextView of the conversation's preview message
	 * @param labelUnread The TextView of the conversation's unread indicator
	 * @param unreadCount The amount of unread messages; set to 0 for none
	 */
	@JvmStatic
	fun bindUnreadStatus(context: Context, labelTitle: TextView, labelMessage: TextView, labelUnread: TextView, unreadCount: Int) {
		if(unreadCount > 0) {
			labelTitle.setTypeface(null, Typeface.BOLD)
			labelTitle.setTextColor(context.resources.getColor(R.color.colorPrimary, null))
			
			labelMessage.setTypeface(null, Typeface.BOLD)
			labelMessage.setTextColor(resolveColorAttr(context, android.R.attr.textColorPrimary))
			
			labelUnread.visibility = View.VISIBLE
			labelUnread.text = intToFormattedString(context.resources, unreadCount)
		} else {
			labelTitle.setTypeface(null, Typeface.NORMAL)
			labelTitle.setTextColor(resolveColorAttr(context, android.R.attr.textColorPrimary))
			
			labelMessage.setTypeface(null, Typeface.NORMAL)
			labelMessage.setTextColor(resolveColorAttr(context, android.R.attr.textColorSecondary))
			
			labelUnread.visibility = View.GONE
		}
	}
	
	/**
	 * Binds a conversation's preview section
	 * @param context The context to use
	 * @param labelMessage The TextView of the conversation's preview message
	 * @param labelStatus The TextView of the conversation's status
	 * @param preview The preview item to display
	 */
	@JvmStatic
	fun bindPreview(context: Context, labelMessage: TextView, labelStatus: TextView, preview: ConversationPreview?) {
		if(preview == null) {
			//Set everything to "unknown"
			labelMessage.setText(R.string.part_unknown)
			labelStatus.setText(R.string.part_unknown)
			labelStatus.setTextColor(resolveColorAttr(context, android.R.attr.textColorSecondary))
			
			return
		}
		
		//Setting the message preview
		labelMessage.text = preview.buildString(context)
		
		if(preview is ConversationPreview.Message && preview.isError) {
			//Setting the status to "not sent"
			labelStatus.setText(R.string.message_senderror)
			labelStatus.setTextColor(context.resources.getColor(R.color.colorError, null))
		} else {
			//Setting the status to the preview time
			labelStatus.text = getLastUpdateStatusTime(context, preview.date)
			labelStatus.setTextColor(resolveColorAttr(context, android.R.attr.textColorSecondary))
		}
	}
	
	/**
	 * Binds a conversation's selected indicator
	 * @param mainView The view of the conversation
	 * @param iconGroup A view representing the conversation's group of members
	 * @param selectionIndicator The conversation's selection indicator view
	 * @param isSelected Whether this conversation is selected
	 * @param animate Whether to animate this change
	 */
	@JvmStatic
	fun bindSelectionIndicator(mainView: View, iconGroup: ViewGroup, selectionIndicator: View, selectionTint: View, isSelected: Boolean, animate: Boolean) {
		mainView.isSelected = isSelected
		
		if(animate) {
			if(isSelected) {
				iconGroup.animate().alpha(0f).withEndAction { iconGroup.visibility = View.INVISIBLE }
				selectionIndicator.animate().alpha(1f).withStartAction { selectionIndicator.visibility = View.VISIBLE }
				selectionTint.animate().alpha(1f).withStartAction { selectionTint.visibility = View.VISIBLE }
			} else {
				iconGroup.animate().alpha(1f).withStartAction { iconGroup.visibility = View.VISIBLE }
				selectionIndicator.animate().alpha(0f).withEndAction { selectionIndicator.visibility = View.GONE }
				selectionTint.animate().alpha(0f).withEndAction { selectionTint.visibility = View.GONE }
			}
		} else {
			if(isSelected) {
				iconGroup.visibility = View.INVISIBLE
				iconGroup.alpha = 0f
				
				selectionIndicator.visibility = View.VISIBLE
				selectionIndicator.alpha = 1f
				selectionTint.visibility = View.VISIBLE
				selectionTint.alpha = 1f
			} else {
				iconGroup.visibility = View.VISIBLE
				iconGroup.alpha = 1f
				
				selectionIndicator.visibility = View.GONE
				selectionIndicator.alpha = 0f
				selectionTint.visibility = View.GONE
				selectionTint.alpha = 0f
			}
		}
	}
}